zero_bin/prover/src/lib.rs

```diff
@@ -13,11 +13,22 @@ use proof_gen::proof_types::GeneratedBlockProof;
 use serde::{Deserialize, Serialize};
 use tokio::io::AsyncWriteExt;
 use tokio::sync::mpsc::Receiver;
-use tokio::sync::oneshot;
+use tokio::sync::{oneshot, Semaphore};
 use trace_decoder::{BlockTrace, OtherBlockData};
 use tracing::{error, info};
 use zero_bin_common::fs::generate_block_proof_file_name;
 
+// All the proving tasks run in parallel. For the big block intervals,
+// this leads to a common situation where the very distant future blocks are
+// being proved first. So we create a pool of permits to limit the number of
+// parallel proving block tasks, and they are retrieved in block increasing
+// order. Initially we put 16, may be a reasonable default. We output proof file
+// when the block batch is finished, so this helps with getting the results
+// sooner.
```

Comment:
purely wording:

```suggestion
// All proving tasks are executed concurrently, which can cause issues for large
// block intervals, where distant future blocks may be proven first.
//
// We then create create a pool to limit the number of parallel proving block
// tasks, retrieving new blocks in increasing order when some block proofs are
// complete.
//
// While proving a block interval, we will output proofs corresponding to block
// batches as soon as they are generated.
```

Reply:
I've removed the mention of `Initially we put 16, may be a reasonable default.` simply as it doesn't carry much additional information on its own, but depending on how we want this to look we may want a TODO for either custom configuration or measurements on better pool sizes.